### PR TITLE
Test case automation - CEPH-83594337

### PR DESCRIPTION
**Pull request by chaitanyadommeti**
Added module for test CEPH-83594337
Modified add_image_to_group_and_verify() of group.py to handle namespace 
Added test config for CEPH-83594337 in yaml file

Success log for  CEPH-83594337: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-HN6EMK/Test_group_snapshot,_clone,_rollback_from_pool_namespace_0.log

Success log to check existing tests which utilise the sub  add_image_to_group_and_verify() without 'namespace' dont break:
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-GLA40F/Test_cloning_and_flattening_group_snapshot_0.log

Success log after addressing review comments:
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-IOZ0KY/

### CODE REVIEW

File: ceph/rbd/workflows/group.py [hunk 1]
```diff
@@ -14,7 +14,7 @@ def create_group_and_verify(**kw):
         kw(dict): Key/value pairs that needs to be provided to the installer
             Example::
             Supported keys:
-                pool(str): pool name where group should be created
+                pool(str): pool nane where group should be created
```
Comment (sunilangadi2) [inline ceph/rbd/workflows/group.py:17]: it should be `name`

File: ceph/rbd/workflows/group.py [hunk 2]
```diff
@@ -92,10 +92,12 @@ def add_image_to_group_and_verify(**kw):
...
```
Comment (sunilangadi2) [inline ceph/rbd/workflows/group.py:95]: same here

File: ceph/rbd/workflows/group.py [hunk 3]
```diff
@@ -129,10 +141,31 @@ def add_image_to_group_and_verify(**kw):
...
+        group_ls_kw.update({"pool": pool})
+        group_ls_kw.update({"group": group})
+        if namespace is not None:
+            group_ls_kw.update({"namespace": namespace})
     (g_ls_out, _) = rbd.group.image.list(**group_ls_kw)
-    if f"{pool}/{image}" in g_ls_out:
+
+    if namespace and f"{image_pool}/{namespace}/{image}" in g_ls_out:
+        log.info(
+            f"Image {image} to the group {group} and namespace {namespace} successfully verified"
```
Comment (sunilangadi2) [inline ceph/rbd/workflows/group.py:171]: It should be namespace-based images added successfully to group Is it?

File: suites/squid/rbd/tier-2_rbd_group.yaml [hunk 4]
```diff
@@ -148,6 +148,23 @@ tests:
...
+        snap: group_snap
+        rep_pool_config:
+          num_pools: 1
+          do_not_create_image: true
+        ec_pool_config:
+          num_pools: 1
+          do_not_create_image: true
+        fio:
+          size: 100M
+      module: test_rbd_group_snapshot_on_pool_namespace
```
Comment (sunilangadi2) [inline suites/squid/rbd/tier-2_rbd_group.yaml:164]: module name is wrong. `.py` is missing

File: tests/rbd/test_rbd_group_snapshot_on_pool_namespace.py [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+        )
+
+        if kw.get("client_node"):
+            client = get_node_by_id(kw.get("ceph_cluster"), kw.get("client_node"))
+        else:
+            client = kw.get("ceph_cluster").get_nodes(role="client")[0]
+        rbd_obj = initial_rbd_config(**kw)
+        pool_types = rbd_obj.get("pool_types")
+        if rbd_obj:
+            log.info("Executing test on Replication and EC pool")
```
Comment (sunilangadi2) [inline tests/rbd/test_rbd_group_snapshot_on_pool_namespace.py:267]: It should be `Replicated`

File: ceph/rbd/workflows/group.py [hunk 6]
```diff
@@ -236,65 +269,3 @@ def rollback_to_snap(**kw):
     else:
         log.info(f"SUCCESS: Group {group} rollbacked to {snap} successfully")
         return 0
-
```
Comment (sunilangadi2) [inline ceph/rbd/workflows/group.py:239]: why it's removed?

File: tests/rbd/test_rbd_group_snapshot_on_pool_namespace.py [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+from ceph.utils import get_node_by_id
+from cli.rbd.rbd import Rbd
+from utility.log import Log
+
+log = Log(__name__)
+
+
+def clone_rollback_group_snapshot_on_namespace(rbd_obj, client, **kw):
+    """
+        Test to verify Cloning and Flattening a Group Snapshot
```
Comment (sunilangadi2) [inline tests/rbd/test_rbd_group_snapshot_on_pool_namespace.py:55]: please correct the method description.

File: tests/rbd/test_rbd_group_snapshot_on_pool_namespace.py [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+    for pool_type in rbd_obj.get("pool_types"):
+        rbd_config = kw.get("config", {}).get(pool_type, {})
+        multi_pool_config = getdict(rbd_config)
+
+        for pool, pool_config in multi_pool_config.items():
+            if "data_pool" in pool_config.keys():
+                _ = pool_config.pop("data_pool")
+
+            kw["pool-name"] = pool
+            namespace = "ns_" + random_string(len=5)
```
Comment (sunilangadi2) [inline tests/rbd/test_rbd_group_snapshot_on_pool_namespace.py:74]: keep `ns` more descriptive as `namespace`

File: tests/rbd/test_rbd_group_snapshot_on_pool_namespace.py [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+                out, err = rbd.create(**create_kw)
+                if err:
+                    log.error(
+                        f"Image {pool}/{namespace}/{image} failed with error {err}"
+                    )
+                    return 1
+                else:
+                    log.info(f"Image {pool}/{namespace}/{image} creation is complete")
+
+            # dev_path = rbd.map(**{"image-or-snap-spec":f"{pool}/{namespace}/{images[0]}"})
```
Comment (sunilangadi2) [inline tests/rbd/test_rbd_group_snapshot_on_pool_namespace.py:94]: why it's commented?

File: tests/rbd/test_rbd_group_snapshot_on_pool_namespace.py [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+            bench_kw = {
```
Comment (sunilangadi2) [inline tests/rbd/test_rbd_group_snapshot_on_pool_namespace.py:95]: if possbile can we take this IO pattern from test config.

File: tests/rbd/test_rbd_group_snapshot_on_pool_namespace.py [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+
+            md5_before_snap = get_md5sum_rbd_image(
+                image_spec=f"{pool}/{namespace}/{images[0]}",
+                rbd=rbd,
+                client=client,
+                file_path="file" + random_string(len=5),
+            )
+            log.info(f"md5 before snap is {md5_before_snap}")
+
+            # 4. Create a group and add the image to an RBD group
```
Comment (sunilangadi2) [inline tests/rbd/test_rbd_group_snapshot_on_pool_namespace.py:120]: why number 4 in this comment?

**Review by sunilangadi2 — CHANGES_REQUESTED** [review]

File: ceph/rbd/workflows/group.py [hunk 12]
```diff
@@ -129,10 +157,22 @@ def add_image_to_group_and_verify(**kw):
...
+        group_ls_kw.update({"pool": pool})
+        group_ls_kw.update({"group": group})
+        if namespace is not None:
+            group_ls_kw.update({"namespace": namespace})
     (g_ls_out, _) = rbd.group.image.list(**group_ls_kw)
-    if f"{pool}/{image}" in g_ls_out:
+
+    if namespace and f"{image_pool}/{namespace}/{image}" in g_ls_out:
+        log.info(f"Added Namespace image in the group {group}  successfully verified")
+        return 0
```
Comment (sunilangadi2) [inline ceph/rbd/workflows/group.py:171]: keep this `return 0` only at the end not needed for each if.

**Review by sunilangadi2 — COMMENTED** [review]
other changes looks good.

**Comment by sunilangadi2** [comment]
EC related data pool not getting deleted here
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-IOZ0KY/Test_group_snapshot,_clone,_rollback_from_pool_namespace_0.log
please add cleanup part similar to this 
https://github.com/red-hat-storage/cephci/pull/4251/files#

**Comment by chaitanyadommeti** [comment]
> EC related data pool not getting deleted here http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-IOZ0KY/Test_group_snapshot,_clone,_rollback_from_pool_namespace_0.log please add cleanup part similar to this https://github.com/red-hat-storage/cephci/pull/4251/files#

Updated script to avoid 'data_pool' key deletion which is resulting in data pool not getting deleted. 
Success log: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-30U1KM/

**Review by sunilangadi2 — CHANGES_REQUESTED** [review]
@chaitanyadommeti please squash all 20 commits to one commit to merge.

**Review by aarsharm — APPROVED** [review]

**Pull request closed without merging**